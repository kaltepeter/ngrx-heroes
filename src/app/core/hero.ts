export class Hero {
  id: number;
  name: string;
}


// for testing
export const generateHero = (idOverride?: number): Hero => ({
  id: idOverride || (Math.floor(Math.random() * 100) + 1),
  name: 'Test name'
});

export const generateHeroArray = (count = 10): Hero[] =>
  // Overwrite random id generation to prevent duplicate IDs:
  Array.apply(null, Array(count)).map((value, index) => generateHero(index + 1));

export const generateHeroMap = (
  heroArray: Array<Hero> = generateHeroArray()
): { ids: Array<number>, entities: any } => ({
  entities: heroArray.reduce(
    (heroMap, hero) => ({ ...heroMap, [hero.id]: hero }),
    {}
  ),
  ids: heroArray.map(hero => hero.id)
});

