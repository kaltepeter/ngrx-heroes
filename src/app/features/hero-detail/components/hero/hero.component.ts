import { Component, OnInit, Input } from '@angular/core';
import { Hero } from '../../../../core/hero';

@Component({
  selector: 'app-hero',
  templateUrl: './hero.component.html',
  styleUrls: ['./hero.component.scss']
})
export class HeroComponent {
  @Input() hero: Hero;
}
