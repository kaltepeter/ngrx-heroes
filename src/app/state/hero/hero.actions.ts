import { Action } from '@ngrx/store';
import { Hero } from '../../core/hero';

export enum HeroActionTypes {
  SearchAllHeroEntities = '[Hero] Search',
  SearchAllHeroEntitiesSuccess = '[Hero] Search Success',
  SearchAllHeroEntitiesFail = '[Hero] Search Fail',

  LoadHeroById = '[Hero] Load By ID',
  LoadHeroByIdSuccess = '[Hero] Load Success',
  LoadHeroByIdFail = '[Hero] Load Fail',
}

// ========================================= SEARCH

export class SearchAllHeroEntities implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntities;
}

export class SearchAllHeroEntitiesSuccess implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntitiesSuccess;
  constructor(public payload: { result: Array<Hero> }) {}
}

export class SearchAllHeroEntitiesFail implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntitiesFail;
  constructor(public payload: { error: string }) {}
}

// ========================================= LOAD BY ID

export class LoadHeroById implements Action {
  readonly type = HeroActionTypes.LoadHeroById;
  constructor(public payload: { id: number }) {}
}

export class LoadHeroByIdSuccess implements Action {
  readonly type = HeroActionTypes.LoadHeroByIdSuccess;
  constructor(public payload: { result: Hero }) {}
}

export class LoadHeroByIdFail implements Action {
  readonly type = HeroActionTypes.LoadHeroByIdFail;
  constructor(public payload: { error: string }) {}
}

export type HeroActions =
  | SearchAllHeroEntities
  | SearchAllHeroEntitiesSuccess
  | SearchAllHeroEntitiesFail
  | LoadHeroById
  | LoadHeroByIdSuccess
  | LoadHeroByIdFail;
