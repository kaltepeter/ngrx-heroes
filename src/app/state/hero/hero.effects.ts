import { Injectable } from '@angular/core';

import { Observable, of } from 'rxjs';
import {
  exhaustMap,
  map,
  catchError,
  switchMap
} from 'rxjs/operators';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';

import {
  HeroActionTypes,
  SearchAllHeroEntities,
  SearchAllHeroEntitiesSuccess,
  SearchAllHeroEntitiesFail,
  LoadHeroById,
  LoadHeroByIdSuccess,
  LoadHeroByIdFail
} from './hero.actions';
import { Hero } from '../../core/hero';
import { HeroService } from '../../core/hero.service';

@Injectable()
export class HeroEffects {
  @Effect()
  search: Observable<Action> = this.actions$.pipe(
    ofType<SearchAllHeroEntities>(HeroActionTypes.SearchAllHeroEntities),
    exhaustMap(() =>
      this.service.getHeroes().pipe(
        map(
          (entities: Array<Hero>) =>
            new SearchAllHeroEntitiesSuccess({ result: entities })
        ),
        catchError(({ message }) =>
          of(new SearchAllHeroEntitiesFail({ error: message }))
        )
      )
    )
  );

  @Effect()
  loadById: Observable<Action> = this.actions$.pipe(
    ofType<LoadHeroById>(HeroActionTypes.LoadHeroById),
    switchMap(action =>
      this.service.getHero(action.payload.id).pipe(
        map((hero: Hero) => new LoadHeroByIdSuccess({ result: hero })),
        catchError(({ message }) =>
          of(new LoadHeroByIdFail({ error: message }))
        )
      )
    )
  );

  constructor(private actions$: Actions, private service: HeroService) {}
}
