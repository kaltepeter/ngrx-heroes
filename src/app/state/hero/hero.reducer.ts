import { Hero } from '../../core/hero';

import * as heroActions from './hero.actions';

export interface HeroState {
  heroes: Hero[];
  hero: Hero;
}

export const initialState: HeroState = {
  heroes: [],
  hero: null
};

export function heroReducer(
  state = initialState,
  action: heroActions.HeroActions
): HeroState {
  switch (action.type) {
    case heroActions.HeroActionTypes.SearchAllHeroEntitiesSuccess: {
      return {
        hero: null,
        heroes: action.payload.result
      };
    }

    case heroActions.HeroActionTypes.LoadHeroByIdSuccess: {
      return {
        heroes: [],
        hero: action.payload.result
      };
    }

    default: {
      return state;
    }
  }
}
