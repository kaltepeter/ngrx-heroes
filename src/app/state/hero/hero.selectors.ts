import { createFeatureSelector, createSelector } from '@ngrx/store';
import * as fromHero from './hero.reducer';

export const selectHeroState = createFeatureSelector<fromHero.State>(
  fromHero.heroFeatureKey
);
