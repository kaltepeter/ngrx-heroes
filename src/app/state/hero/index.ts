import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromHeroState from './hero.reducer';
import {HeroState} from './hero.reducer';
export const getHeroState = createFeatureSelector<HeroState>('hero');

export const {
  selectIds: heroIds,
  selectEntities: entities,
  selectAll: heroes,
  selectTotal: heroCount
} = fromHeroState.adapter.getSelectors(getHeroState);

export const topHeroes = createSelector(
  heroes,
  data => data.slice(1, 5)
);

export const selectedHeroId = createSelector(
  getHeroState,
  fromHeroState.getSelectedId
);
export const selectedHero = createSelector(
  selectedHeroId,
  entities,
  (heroId, heroEntities) => heroId && heroEntities[heroId]
);

export const heroLoading = createSelector(
  getHeroState,
  fromHeroState.getLoading
);

export const heroError = createSelector(
  getHeroState,
  fromHeroState.getError
);
