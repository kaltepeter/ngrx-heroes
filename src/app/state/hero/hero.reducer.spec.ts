import {
  Hero,
  generateHero,
  generateHeroArray,
  generateHeroMap
} from '../../core/hero';
import * as actions from './hero.actions';
import {
  heroReducer,
  initialHeroState,
  getSelectedId,
  getLoading,
  getError
} from './hero.reducer';
import { Update } from '@ngrx/entity';

const INITIAL_STATE_WITH_ERROR = {
  ...initialHeroState,
  error: 'some error'
};
const BLANK_ERROR_MESSAGE = '';

describe('heroReducer', () => {
  describe('upon an undefined action', () => {
    it('should return the default state upon an undefined action', () => {
      const action = { type: 'NOT DEFINED' } as any;

      expect(heroReducer(initialHeroState, action)).toEqual(initialHeroState);
    });
  });

  describe('upon SearchAllHeroEntities', () => {
    it('should remove Hero entities, set loading to true, and clear any error', () => {
      const initialHeroStateWithHeroEntities = {
        ...INITIAL_STATE_WITH_ERROR,
        ...generateHeroMap()
      };
      const action = new actions.SearchAllHeroEntities();

      expect(heroReducer(initialHeroStateWithHeroEntities, action)).toMatchSnapshot();
    });
  });

  describe('upon SearchAllHeroEntitiesSuccess', () => {
    it('should add Hero entities, set loading to false, and clear any error', () => {
      const result = generateHeroArray();
      const action = new actions.SearchAllHeroEntitiesSuccess({ result });

      expect(heroReducer(INITIAL_STATE_WITH_ERROR, action)).toMatchSnapshot();
    });
  });

  describe('upon SearchAllHeroEntitiesFail', () => {
    it('should set loading to false and echo the error', () => {
      const error = 'test search error';
      const action = new actions.SearchAllHeroEntitiesFail({ error });

      expect(heroReducer(initialHeroState, action)).toMatchSnapshot();
    });
  });

  describe('upon LoadHeroById', () => {
    it('should remove hero entities, set selected id, and clear any error', () => {
      const id = 8675309;
      const initialHeroStateWithHeroEntities = {
        ...INITIAL_STATE_WITH_ERROR,
        ...generateHeroMap()
      };
      const action = new actions.LoadHeroById({ id });

      expect(heroReducer(initialHeroStateWithHeroEntities, action)).toMatchSnapshot();
    });
  });

  describe('upon LoadHeroByIdSuccess', () => {
    it('should add the given Hero, set loading to false, and clear any error', () => {
      const result = generateHero(54);
      const action = new actions.LoadHeroByIdSuccess({ result });

      expect(heroReducer(INITIAL_STATE_WITH_ERROR, action)).toMatchSnapshot();
    });
  });

  describe('upon LoadHeroByIdFail', () => {
    it('should set loading to false and echo the error', () => {
      const error = 'test load by id error';
      const action = new actions.LoadHeroByIdFail({ error });

      expect(heroReducer(initialHeroState, action)).toMatchSnapshot();
    });
  });
});

describe('getters', () => {
  describe('getSelectedId', () => {
    it('should return the selected id', () => {
      expect(getSelectedId(initialHeroState)).toEqual(initialHeroState.selectedId);
    });
  });
  describe('getLoading', () => {
    it('should return the selected id', () => {
      expect(getLoading(initialHeroState)).toEqual(initialHeroState.loading);
    });
  });
  describe('getError', () => {
    it('should return the selected id', () => {
      expect(getError(INITIAL_STATE_WITH_ERROR))
        .toEqual(INITIAL_STATE_WITH_ERROR.error);
    });
  });
});
