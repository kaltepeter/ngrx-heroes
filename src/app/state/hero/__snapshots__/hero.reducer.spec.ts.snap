// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heroReducer upon LoadHeroById should remove hero entities, set selected id, and clear any error 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loading": true,
  "selectedId": 8675309,
}
`;

exports[`heroReducer upon LoadHeroByIdFail should set loading to false and echo the error 1`] = `
Object {
  "entities": Object {},
  "error": "Hero load failed: test load by id error",
  "ids": Array [],
  "loading": false,
  "selectedId": null,
}
`;

exports[`heroReducer upon LoadHeroByIdSuccess should add the given Hero, set loading to false, and clear any error 1`] = `
Object {
  "entities": Object {
    "54": Object {
      "id": 54,
      "name": "Test name",
    },
  },
  "error": "",
  "ids": Array [
    54,
  ],
  "loading": false,
  "selectedId": null,
}
`;

exports[`heroReducer upon SearchAllHeroEntities should remove Hero entities, set loading to true, and clear any error 1`] = `
Object {
  "entities": Object {},
  "error": "",
  "ids": Array [],
  "loading": true,
  "selectedId": null,
}
`;

exports[`heroReducer upon SearchAllHeroEntitiesFail should set loading to false and echo the error 1`] = `
Object {
  "entities": Object {},
  "error": "Hero search failed: test search error",
  "ids": Array [],
  "loading": false,
  "selectedId": null,
}
`;

exports[`heroReducer upon SearchAllHeroEntitiesSuccess should add Hero entities, set loading to false, and clear any error 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Test name",
    },
    "10": Object {
      "id": 10,
      "name": "Test name",
    },
    "2": Object {
      "id": 2,
      "name": "Test name",
    },
    "3": Object {
      "id": 3,
      "name": "Test name",
    },
    "4": Object {
      "id": 4,
      "name": "Test name",
    },
    "5": Object {
      "id": 5,
      "name": "Test name",
    },
    "6": Object {
      "id": 6,
      "name": "Test name",
    },
    "7": Object {
      "id": 7,
      "name": "Test name",
    },
    "8": Object {
      "id": 8,
      "name": "Test name",
    },
    "9": Object {
      "id": 9,
      "name": "Test name",
    },
  },
  "error": "",
  "ids": Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  "loading": false,
  "selectedId": null,
}
`;
